### DISCUSSION TITLE
github not recognizing lgpl

### DISCUSSION THREAD

**Discussion by kindnesstheorist** [community/community#22258] [New to GitHub]
<p>The way the LGPL works is that it modifies the GPL. Therefore a program “released under the LGPL” in fact is a program whose license consists of the GPL and the LGPL. Conventionally this is represented by two files named COPYING and COPYING.LESSER, respectively. Including both licenses is not a source of ambiguity; the licenses are written such that the LGPL takes precedence when both are present. Indeed, <a href="http://gnu.org" rel="nofollow noopener">gnu.org</a> advises against shipping the LGPL <em>without</em> including a copy of the GPL. (<a href="http://www.gnu.org/licenses/gpl-howto.html" rel="nofollow noopener">source</a>)</p>
<p>TL;DR, according to <a href="http://gnu.org" rel="nofollow noopener">gnu.org</a>, “LGPL program” = program + GPL + LGPL.</p>
<p>Now, GitHub’s license-managing software, it doesn’t think like that. It thinks something like this:</p>
<pre><code>function autoDetectLicense(repository, licenses)
for file in repository {
    if !file.isDirectory() {
for license in licenses {
            if similarEnough(file.getContentsAsText(), license.text) return license.name
}
}
return "licensenotdetected"
}
</code></pre>
<p>(Assume “licenses” is an array that starts with “gpl, lgpl” and contains all the licenses GitHub recognizes.)</p>
<p>This isn’t an especially unreasonable way of doing things, but it fails pretty badly on the LGPL. It could be patched by changing the order of the array to put the LGPL before the GPL (assuming it’s an <em>array</em> and not, say, a hash table), but “the seemingly arbitrary order of this array is actually important to functioning in a way that not only could or will but <em>has</em> come up” is no way to do programming.</p>
<p>I see several potential patches that don’t do that:</p>
<ol>
<li>
<p>Collect all license files with recognizable contents from the repository’s top-level folder. Then, if there’s exactly one, auto-detect that one. If there’s more than one, let the user decide which one to display.</p>
</li>
<li>
<p>Like 1, but allow simultaneous publication under multiple licenses.</p>
</li>
<li>
<p>Like the current behavior, but treat the GPL, specifically, as a potentially incomplete license, and keep looking until either the LGPL is found or the LGPL is definitely not in the top-level directory. If the former, LGPL; if the latter; GPL.</p>
</li>
<li>
<p>Like the current behavior, but somewhere in the settings, allow the user to manually point the auto-detector at a particular license file (instead of having it always detect the first one in its sorting order).</p>
</li>
</ol>
<p>For the record, I like 2 best, but 1 or 4 would probably be easier to implement.</p>
<p>Example of the issue in action: <a href="https://github.com/kindnesstheorist/lgpl-java-template" rel="nofollow noopener">https://github.com/kindnesstheorist/lgpl-java-template</a>.</p>
<p>Is this a known issue? Anybody have a workaround? I don’t want people to scroll past my code because they have the wrong idea about its license, but I also really don’t want to use the LGPL in a way that its publishers specifically say not to do.</p>

**Comment by lee-dohm** [comment, marked as answer]
<p>GitHub uses the <a href="https://github.com/licensee/licensee" rel="nofollow noopener">licensee project</a> to do its license detection. It appears they already have an <a href="https://github.com/licensee/licensee/issues/287" rel="nofollow noopener">issue open for LGPL detection</a>. You may want to follow along there.</p>
<p>I hope that helps!</p>

**Comment by kindnesstheorist** [comment]
<p>…huh. How’d this get marked as a solution before I saw it?</p>
<p>Anyway, brb, off to rename my license files and see if that fixes it.</p>

**Comment by kindnesstheorist** [comment]
<p>As soon as I renamed COPYING.LESSER.txt to COPYING.LESSER, it recognized the LGPL. Shouldn’t have tried to get fancy, I guess. Thanks for the info; the discussion was interesting to read and I’m glad this is getting fixed soon.</p>
